great_expectations/types/base.py

```diff
@@ -27,138 +27,83 @@ def __dir__(self):
     def __deepcopy__(self, memo):
         return DotDict([(copy.deepcopy(k, memo), copy.deepcopy(v, memo)) for k, v in self.items()])
 
+    # The following are required to support yaml serialization, since we do not raise
+    # AttributeError from __getattr__ in DotDict. We *do* raise that AttributeError when it is possible to know
+    # a given attribute is not allowed (because it's not in _allowed_keys)
+    _yaml_merge = []
 
-# Inspiration : https://codereview.stackexchange.com/questions/81794/dictionary-with-restricted-keys
-class LooselyTypedDotDict(DotDict):
-    """dot.notation access to dictionary attributes, with limited keys
-
-
-    Note: this class is pretty useless on its own.
-    You need to subclass it like so:
+    @classmethod
+    def yaml_anchor(cls):
+        # This is required since our dotdict allows *any* access via dotNotation, blocking the normal
+        # behavior of raising an AttributeError when trying to access a nonexistent function
+        return None
 
-    class MyLooselyTypedDotDict(LooselyTypedDotDict):
-        _allowed_keys = set([
-            "x", "y", "z"
-        ])
+    @classmethod
+    def to_yaml(cls, representer, node):
```

Comment:
❤️ ❤️ ❤️ 
This is going to make storing configs, etc. go so much more smoothly.

Reply:
Thanks. One other thing this PR does that may not have been obvious is to add *args to all of the constructors, so that the entire chain of these classes now follows the dict semantics, so they are close-to-faithful classes that give the option of all the desired serializations.

A further thing we might want to do is to *require* that everything in them is json serializable...